Prepend a version number too in source tarballs

mortenpi opened the pull request:
Currently the files in _source_ tarballs just have `julia/` prepended to them (e.g. [1.0.1 tarball](https://github.com/JuliaLang/julia/releases/download/v1.0.1/julia-1.0.1.tar.gz)), but I propose including the version number too, i.e. prepend `julia-x.y.z/`.

I switched over to using the `--transform` option of `tar` for modifying the filenames. However, this might not be very portable -- it seems to be that `--transform`/`--xform` is specific to GNU tar. [BSD tar](https://www.freebsd.org/cgi/man.cgi?query=tar&apropos=0&sektion=1&manpath=FreeBSD+11.2-RELEASE+and+Ports&arch=default&format=html) does not appear to have it.. or rather calls a similar option `-s` if I understand correctly.

How important is it to keep these targets cross-platform? I do not have any other idea how to approach this at the moment, other than doing something similar to the original implementation of `cd`-ing up and then renaming or symlinking the directory, or copying files over to temporary directory.

mortenpi committed:
Also prepend a version number in source tarballs

ararslan commented:
> How important is it to keep these targets cross-platform?

It is required. These targets should be able to be run on any system that can build Julia.

mortenpi committed:
Take 2: use symlinks

Instead of transforming the file names, just create a temporary symlink
in the source directory that points back to the root of the source
directory. As long as symlinks work properly, this should be a portable
solution.

mortenpi commented:
Different approach now: as we are using an explicit file list, it actually shouldn't be a problem to just temporarily create a symlink in the source tree which points back to the root. I would hope that this would work properly on all UNIX-y platforms. Less certain about Windows though.

mortenpi commented:
I checked that this works as intended on Windows in a Cygwin build. And by interpolation, I hope that it works on everything in between.

staticfloat commented on Makefile:
```diff
@@ -515,10 +515,14 @@ endif
+	DIRNAME=julia-$(JULIA_COMMIT); \
 	sed -e "s_.*_$$DIRNAME/&_" light-source-dist.tmp > light-source-dist.tmp1; \
-	cd ../ && tar -cz --no-recursion -T $$DIRNAME/light-source-dist.tmp1 -f $$DIRNAME/julia-$(JULIA_VERSION)_$(JULIA_COMMIT).tar.gz
+	ln -s . $$DIRNAME || exit 1; \
```
Why `|| exit 1`?   I'm not sure I see why that would change any outcome?

mortenpi replied:
I always assumed that just `exit` would just return `0`, but it turns out that is not the case -- it returns `$?`. Removed the `1`s.

staticfloat replied:
`exit 1` explicitly causes `bash` to exit with return code `1`, so I think what you want is `exit 0`.

staticfloat replied:
Actually, I'm not sure that `exit 0` is what you want.  What are you trying to do?  Do you want to ignore `ln -s` errors?  Or do you want `ln -s` errors to prevent running the `tar` command?  Or do you want `ln -s` errors to run the `tar` command but prevent the `make` build from finishing?

staticfloat replied:
I think the best thing to do then is to not have an `exit` at all, and replace `ln -s . || exit 1; \` with `ln -s . && \`

mortenpi replied:
I also need to connect `tar` and `rm` with `&&` then. And could connect `sed` too -- unlikely to fail, but might as well check.

vtjnash replied:
It’s a Makefile—it’s already checking the exit code of each bash process (unless prefixed with -). No need to duplicate that.

vtjnash replied:
Also, `|| exit` is the same as `|| true`, since the default exit code is success

staticfloat commented:
I'm actually quite surprised this worked on Windows Cygwin; are you on a recent Windows 10?  I believe that anything older than a recent Windows 10 might require admin privileges to create symlinks.  (I'm not that worried about it since this is an unusual operation, so as long as it works for "some" windows machines I'm okay with it)

mortenpi committed:
Remove status from exit calls

mortenpi commented:
It should be latest -- it is a fresh install from a few days ago on a VM with all the post-install updates (says version 10.0.17134).

But in any case, doesn't Cygwin implement symlinks as normals files from the Windows perspective? I am basing this on [this SO answer](https://stackoverflow.com/a/18659632/1601695), and when I tried it out and looked at the symlink in the Windows file manager, it indeed appears to be just a normal file.

mortenpi commented:
AV failure seems unrelated.

staticfloat commented:
> But in any case, doesn't Cygwin implement symlinks as normal files from the Windows perspective?

I didn't realize this; that's great.

mortenpi committed:
Split into separate commands

mortenpi commented:
Friendly bump -- anything left to do here?

StefanKarpinski commented:
Seems fine to me but I'm not sure I'm qualified to review. Who would you like to review?

mortenpi commented:
Not sure. Maybe @staticfloat or @vtjnash could take another look?

staticfloat commented:
Looks good to me.  Thanks @mortenpi

staticfloat commented:
.....ah I clicked "rebase and merge", not "squash and merge".  🙄
